### PR TITLE
Port over the NixOS fish module

### PR DESCRIPTION
**Pull request by cbarrett**
This PR essentially replaces this file with the nixos fish module. The only difference between the two now is nuking the reference to `environment.shells`  🎉 

The NixOS fish module has a number of nice features that we're gaining, such as more customizable initialization.

### CODE REVIEW

File: modules/programs/fish.nix [hunk 1]
```diff
@@ -45,109 +21,163 @@ in
...
+    environment.etc."fish/foreign-env/interactiveShellInit".text = cfge.interactiveShellInit;
+    environment.etc."fish/foreign-env/extraInit".text= cfge.extraInit;
+
+    environment.etc."fish/nixos-env-preinit.fish".text = ''
+      # This happens before $__fish_datadir/config.fish sets fish_function_path, so it is currently
+      # unset. We set it and then completely erase it, leaving its configuration to $__fish_datadir/config.fish
+      set fish_function_path ${pkgs.fish-foreign-env}/share/fish-foreign-env/functions $__fish_datadir/functions
+     
+      # For "reasons not remembered," setEnvironment does not set PATH, so do it here
+      set PATH ${replaceStrings [":"] [" "] config.environment.systemPath}
```
Comment (LnL7) [inline modules/programs/fish.nix]: I think it was something related to ordering, but I guess it's actually useful in this case. How is the separator handled on nixos?

File: modules/programs/fish.nix [hunk 2]
```diff
@@ -45,109 +21,163 @@ in
...
+    environment.etc."fish/nixos-env-preinit.fish".text = ''
+      # This happens before $__fish_datadir/config.fish sets fish_function_path, so it is currently
+      # unset. We set it and then completely erase it, leaving its configuration to $__fish_datadir/config.fish
+      set fish_function_path ${pkgs.fish-foreign-env}/share/fish-foreign-env/functions $__fish_datadir/functions
+     
+      # For "reasons not remembered," setEnvironment does not set PATH, so do it here
+      set PATH ${replaceStrings [":"] [" "] config.environment.systemPath} 
+
+      # source the NixOS environment config
+      fenv source ${config.system.build.setEnvironment}
```
Comment (LnL7) [inline modules/programs/fish.nix]: I inlined this in the other modules since nothing actually needed a separate file. But I can change this to a `writeText` and use the text attribute instead.

Comment (LnL7) [reply modules/programs/fish.nix]: See a8b47702751e5f53448c65ba1655747a88714445 and e8355cdd938c85b86024d22bd44ed7e3c939e17b, that should resolve most of differences with the nixos modules.

**Comment by yurrriq** [comment]
Thanks for this. I’m reading the diff on my phone and it looks like you’ve done away with `variables` in the config. I use that in system config. Is there another approach now?

File: modules/programs/fish.nix (hunk 1 above)
Comment (therealpxc) [reply modules/programs/fish.nix]: Since setting the PATH happens inside `fenv`'s temporary bash environment, the foreign-env plugin for _Oh My Fish!_ handles [converting the separator for us](https://github.com/oh-my-fish/plugin-foreign-env/blob/master/functions/fenv.apply.fish#L31). This applies just as well to nix-darwin since we're also using fenv. :-)

File: modules/programs/fish.nix [hunk 3]
```diff
@@ -45,109 +21,163 @@ in
...
-      set PATH ${replaceStrings [":"] [" "] config.environment.systemPath}
-
-      ${config.system.build.setEnvironment}
-
-      fenv source ${foreignEnv}
-      ${cfg.shellInit}
-
-      ${concatStringsSep "\n" fishVariables}
+      # if we haven't sourced the general config, do it
+      if not set -q __fish_nixos_general_config_sourced
```
Comment (therealpxc) [inline modules/programs/fish.nix]: @LnL7 @cbarrett 

do we care that references to NixOS have now ended up in the module? Would it be worth adding a PR to Nixpkgs which changes these to generic names, perhaps substituting `nix` for `nixos` and makes a note that this path is depended-on in nix-darwin and potentially other declarative configuration/module systems for Nix and not just NixOS?

Is it worth it just for the name change? Is it worth it to leave a note to Nixpkgs/NixOS developers so they know to at least ping people should that change?

File: modules/programs/fish.nix [hunk 4]
```diff
@@ -45,109 +21,163 @@ in
...
-    environment.pathsToLink = [ "/share/fish" ];
-
-    environment.loginShell = "${shell}/bin/fish -l";
-    environment.variables.SHELL = "${shell}/bin/fish";
+    environment.etc."fish/foreign-env/shellInit".text = cfge.shellInit;
+    environment.etc."fish/foreign-env/loginShellInit".text = cfge.loginShellInit;
+    environment.etc."fish/foreign-env/interactiveShellInit".text = cfge.interactiveShellInit;
+    environment.etc."fish/foreign-env/extraInit".text= cfge.extraInit;
+
+    environment.etc."fish/nixos-env-preinit.fish".text = ''
```
Comment (therealpxc) [inline modules/programs/fish.nix]: In the Fish package in Nixpkgs, Fish looks for this file, `/etc/fish/nixos-env-preinit.fish`. This is fine for us, but it means that we have to 'own' `/etc/fish` in `nix-darwin`. If we changed the Fish package to refer to `/etc/static/fish/nixos-env-preinit.fish` or something like that, this piece (and other pieces, if we changed them accordingly) would work even if `nix-darwin` can't or doesn't write directly to `/etc/fish`. Is that desirable? I don't think it would disrupt the way things work on NixOS.

tl;dr: Can we safely assume that `/etc/static` is where both nix-darwin and NixOS will write Fish config files to, and if so, should we change the modules (on both) to refer to those paths instead of the ones that live under `/etc`?

File: modules/programs/fish.nix (hunk 1 above)
Comment (LnL7) [reply modules/programs/fish.nix]: Ah, interesting. I might change it to just set `environment.variables.PATH` at some point then.

**Comment by therealpxc** [comment]
@yurrriq Good catch. I think maybe it would be best if we could use a global 'variables' attribute in config.environment and make sure that those variable changes are represented in `setEnvironment` or `extraInit`. (There might already be one; I haven't looked.)

I don't think that the foreign-env plugin currently applies aliases, but I think if we could enhance it to do so that would be ideal, because then we could put all of this stuff in generically as POSIX shell scripts and have `fenv` take care of all of it. Then setting variables can, for users, be independent of shell selection.

If this got through as it is now, you could set those variables with export directives inside `shellInit` or any of the other `*ShellInit` attributes in the your nix-darwin configuration.

I don't think you should have to, though. Being able to set aliases and variables directly, without worrying about the logic of when those things get sourced seems like it should be a feature. It looks like it's missing in the NixOS module, so I will seek to update that as well according to what we decide is a good idea here.

**Comment by cbarrett** [comment]
Thanks for all the comments! I'll take a look at addressing them tonight hopefully, sometime this week for sure.

**Comment by therealpxc** [comment]
Hey y'all. I [just submitted a PR for Fish in Nixpkgs](https://github.com/NixOS/nixpkgs/pull/30284) which makes it so that Fish installed via Nix takes care of NIxifying the environment automatically and transparently, regardless of whether you're using Fish on NixOS, Nix with GNU+Linux, Nix on macOS, or with nix-darwin.

It still needs more manual testing and probably other work (I think we should add some tests for initialization behavior as well), but it's ready for your comments. I'm using it with nix-darwin (a version similar to this PR (slightly modified and based on @LnL7 's recent updates) on this laptop and I'll be testing it on macOS without nix-darwin tonight, as well as on NixOS. I may not have a non-NixOS GNU+Linux machine online to test on until tomorrow.

**Comment by cbarrett** [comment]
@therealpxc I agree that better alias support would be good. FWIW as of 2.5.0, fish seems to operate much closer to bash wrt. aliases. (e.g. typing `alias` will list all the aliases).

**Comment by therealpxc** [comment]
So I hadn't got back to this, but the way this is currently set up, does it actually translates all the system-wide aliases correctly so that `fenv` is unnecessary for that part? I think I mistakenly thought the aliases in the module were a Fish-specific thing.

File: modules/programs/fish.nix (hunk 3 above)
Comment (cbarrett) [reply modules/programs/fish.nix]: Good point. I'm down with changing the names to `nix_darwin` instead of `nixos`. How do others feel?

**Comment by cbarrett** [comment]
Ah yeah, I forgot to bring back adding variables (thanks for the heads up @yurrriq)

**Comment by cbarrett** [comment]
Looking into the test failure…

**Comment by yurrriq** [comment]
I’ve amended my config so it’s moot for me, but @therealpxc and I have been working though reinstalling Nix and testing all this cleanly. I’ve borked my work machine now and it’s rebuilding the nix-darwin system manually (curl error 77 x a million) while at this show... here’s hoping it’s all good when I get home / by tomorrow morning..

**Comment by yurrriq** [comment]
*Update*: Both my personal and work machines now have fully-functioning, fresh installs of Nix and nix-darwin and I can test out fish changes on Saturday.

**Comment by therealpxc** [comment]
(If you try to test these changes and my Nixpkgs changes at the same time, your system will be broken. Eventually we will make sure that the changes in Nixpkgs are met with whatever changes are needed in nix-darwin to make things work, but we're not letting one PR hold up the other for now.)

**Comment by LnL7** [comment]
@yurrriq Have you had a chance to test this with your configuration, did anything break?

**Comment by yurrriq** [comment]
Last I checked it didn't break anything. FWIW I stopped using `programs.fish.variables`.

**Comment by cbarrett** [comment]
@LnL7 This should be ready to go now!

**Comment by yurrriq** [comment]
:tada: I’ll try this out again on my personal machine shortly.

**Comment by yurrriq** [comment]
This has potentially introduced an issue on my system:
> set: Tried to change the read-only variable '_'

I'm trying to find the cause, but also doing a `nix-channel --update` now.

**Comment by therealpxc** [comment]
This is a cosmetic issue and a side effect of using fenv with the
setEnvironment script. It's being addressed in our fenv patch in the
fish-foreign-env package in Nixpkgs. I think Colin recently opened a PR for
it.

**Comment by cbarrett** [comment]
Correct. I've been busy with a class and have been neglecting Nix. I aim to make progress on this next week as I'm off from work.
